design-proposals/klip-47-non-null-columns-and-component-types.md

```diff
@@ -0,0 +1,183 @@
+existing mechanism/syntax to indicate the type is not nullable. This leads to some undesirable behavior when users use Schema Registry.
+For example, if a map key is nullable, Schema Registry will interpret the map schema as a list of key/value map entries. Since ksqlDB map keys
+are always nullable, ksqlDB will take schemas with maps and write them to Schema Registry as a list of key/value map entries, which is a confusing behavior for users.
+Having map keys that are `NOT NULL` would allow ksqlDB maps to be represented properly in Schema Registry.
+* https://github.com/confluentinc/ksql/issues/6049
+* https://github.com/confluentinc/ksql/issues/2619
+
+For the rest of the KLIP, I'll refer to any stream/table with a `NOT NULL` column as a `NOT NULL` stream/table.
+
+### Data Ownership
+Since ksqlDB relies on Kafka to store the data it reads/writes, the owner of the data in a topic isn't solely the ksqlDB service.
+Producers outside ksqlDB are able to write into the topic. This means that the underlying topic for a `NOT NULL` stream/table can still
+have null column records in it. There's no way around this as ksqlDB can't prevent outside producers from writing into the Kafka topics.
+
+With that in mind, the `NOT NULL` constraint can be enforced within the ksqlDB service. When reading records from a source that has a `NOT NULL` 
+column constraint, invalid records will be dropped, and an error sent to the processing log. When attempting to write an invalid
+record to a sink, the record would also be dropped, and an error sent to the processing log. A `INSERT INTO VALUES` statement with
+a null column record would error out. There would be clear documentation for users that `NOT NULL` does not prevent invalid records in the 
+Kafka topic due to records from producers outside ksqlDB.
+
+If there are concerns that applying a constraint over existing data (declaring a stream/table over an existing topic) would lead to records getting dropped, 
+we could require a `DEFAULT` value be specified with `NOT NULL`. This is explained later in the KLIP and needs to be discussed further
+ as it's not within scope of the original issue filed.
+
+## What is in scope
+* Allowing users to specify `NOT NULL` for columns
+* Prevent reading records with a null column value from `NOT NULL` streams/tables
+* Prevent writing records with a null column value to `NOT NULL` streams/tables
+* Allow users to specify `NOT NULL` for array type, map key/value type, and struct field types
+* (maybe) Allow users to specify a default value when applying `NOT NULL`
+
+## Design
+This design section is only focused on `NOT NULL` for now until we decide whether `DEFAULT` is required or not.
+
+We can add a new `optional` variable to the `SqlType` class to indicate whether the type is optional or not. A `SqlType` variable is in each `Column` object so 
```

Comment:
The `SqlType` just defines a type. I don't we should add column constraints here. What about the `Column` class? It seems the right class to add different column attributes, like not_null, default_value, etc.

Reply:
Sound like implementation details? Seems we can omit those in the KLIP?